CDAP-5117 Implementation for creating the local dataset instances and using them for MapReduce programs in the Workflow.

sagarkapare opened the pull request:
Following changes are implemented in this PR: 
1. Added implementation for creating and deleting the local datasets in the Workflow driver.
2. Using `ForwardingDatasetFramework` in the `WorkflowDriver` now which delegates the calls to the underlying dataset framework. It also maintains the mapping of the local dataset to the name assigned to them for the particular run of the Workflow which is of the form `datasetName.<workflow_run_id>`.
3. Passing the dataset name mapping to the `MapperWrapper` as well through the job configuration.

Spark related changes and deletion of the local datasets based on the the runtime argument will be done in separate PR.

sagarkapare commented:
Build is running here - http://builds.cask.co/browse/CDAP-DUT3631-2

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/ForwardingDatasetFramework.java:
```diff
@@ -0,0 +1,199 @@
+ * dataset framework. Before forwarding the calls, local dataset names are mapped to the names
+ * assigned to them for the current run of the {@link Workflow}.
+ */
+public class ForwardingDatasetFramework implements DatasetFramework, ProgramContextAware {
```
By convention of forwarding object, the `ForwardingDatasetFramework` shouldn't implements `ProgramContextAware` and shouldn't have logic for dataset name mapping. The class hierarchy should be like this:

  * `FowardingDatasetFramework implments DatasetFramework` - Simply delegate all methods to another instance
  * `WorkflowDatasetFramework extends ForwardingDatasetFramework implements ProgramContextAware` - Do the local dataset name mapping by overriding methods. It also implements the `ProgramContextAware` methods and do forwarding in case the delegate is an instance of `ProgramContextAware`.
  * The existing `LineageWriterDatasetFramework` should extends from `ForwardingDatasetFramework` as well to reduce code duplication.

chtyim replied:
You can also have a `ForwardingProgramContextAwareDatasetFramework` that extends from `ForwardingDatasetFramework` and implements `ProgramContextAware` with forwarding. Then the `WorkflowDatasetFramework` will be extending from `ForwardingProgramContextAwareDatasetFramework`.

sagarkapare replied:
Updated code per suggestions.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -427,14 +446,39 @@ private void executeCondition(ApplicationSpecification appSpec, WorkflowConditio
+      String localInstanceName = getLocalDatasetName(instanceEntry.getKey(), runId.getId());
+      Id.DatasetInstance instanceId = Id.DatasetInstance.from(program.getNamespaceId(), localInstanceName);
+      DatasetCreationSpec instanceSpec = instanceEntry.getValue();
+      LOG.info("Adding Workflow local dataset instance: {}", localInstanceName);
```
Sounds more like a debug log.

sagarkapare replied:
Since it only gets printed once I thought its better to have it as `INFO` so that we know what local datasets are getting created. However realized that dataset framework also prints the INFO log for the same. Moved to `DEBUG`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -427,14 +446,39 @@ private void executeCondition(ApplicationSpecification appSpec, WorkflowConditio
+    for (Map.Entry<String, DatasetCreationSpec> instanceEntry : workflowSpec.getLocalDatasetSpecs().entrySet()) {
+      String localInstanceName = getLocalDatasetName(instanceEntry.getKey(), runId.getId());
+      Id.DatasetInstance instanceId = Id.DatasetInstance.from(program.getNamespaceId(), localInstanceName);
+      LOG.info("Deleting Workflow local dataset instance: {}", localInstanceName);
```
Also, more like a debug log.

sagarkapare replied:
Moved to `DEBUG`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -427,14 +446,39 @@ private void executeCondition(ApplicationSpecification appSpec, WorkflowConditio
+      LOG.info("Deleting Workflow local dataset instance: {}", localInstanceName);
+      try {
+        datasetFramework.deleteInstance(instanceId);
+      } catch (IOException | DatasetManagementException e) {
```
Shouldn't you catch Throwable? No matter what error, we should try our best to delete all local dataset.

sagarkapare replied:
Yes. Agree. Catching Throwable.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -427,14 +446,39 @@ private void executeCondition(ApplicationSpecification appSpec, WorkflowConditio
                new InstantiatorFactory(false), program.getClassLoader(), token);
 
     LOG.info("Workflow execution succeeded for {}", workflowSpec);
+    deleteLocalDatasets();
```
So the contract is if a workflow execution fail, local dataset will never get deleted?

sagarkapare replied:
Added the call in the `finally` block.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -427,14 +446,39 @@ private void executeCondition(ApplicationSpecification appSpec, WorkflowConditio
+  }
+
+  private void deleteLocalDatasets() {
+    for (Map.Entry<String, DatasetCreationSpec> instanceEntry : workflowSpec.getLocalDatasetSpecs().entrySet()) {
```
Isn't that we have some runtime arguments to check to decide if a particular local dataset need to be retained or not?

sagarkapare replied:
Yes. Will add that as a separate PR for handling the runtime argument under the JIRA https://issues.cask.co/browse/CDAP-5118

chtyim commented on cdap-unit-test/src/test/java/co/cask/cdap/test/app/WorkflowAppWithLocalDatasets.java:
```diff
@@ -0,0 +1,165 @@
+        waitFile.createNewFile();
+
+        int uniqueWordCount = 0;
+        CloseableIterator<KeyValue<byte[], byte[]>> scanner = wordCount.scan(null, null);
```
The scanner isn't closed after being used.

sagarkapare replied:
oh yes. fixed.

chtyim commented on cdap-unit-test/src/test/java/co/cask/cdap/test/app/WorkflowAppWithLocalDatasets.java:
```diff
@@ -0,0 +1,165 @@
+    private Text word = new Text();
+
+    public void map(Object key, Text value, Context context
+    ) throws IOException, InterruptedException {
```
Wrong place for new line.

chtyim commented on cdap-unit-test/src/test/java/co/cask/cdap/test/app/WorkflowAppWithLocalDatasets.java:
```diff
@@ -0,0 +1,165 @@
+      setDescription("Workflow program with local datasets.");
+      createLocalDataset(WORDCOUNT_DATASET, KeyValueTable.class);
+      addMapReduce("WordCount");
+      addAction(new LocalDatasetReader());
```
Can you add a spark stage?

sagarkapare replied:
Wanted to get feedback on the MapReduce programs first. As mentioned in the PR description will add the Spark related changes in the next PR.

chtyim commented on cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestFrameworkTestRun.java:
```diff
@@ -439,6 +440,61 @@ public void testCustomActionDatasetAccess() throws Exception {
+    // Write some data to the input file
+    File inputDir = tmpFolder.newFolder();
+
+    File inputFile = new File(inputDir.getPath() + "/words.txt");
```
Better use the `File(File parent, String path)` constructor: `new File(inputDir, "words.txt")`

sagarkapare replied:
Updated code to use the `File(File parent, String path)` constructor.

chtyim commented on cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestFrameworkTestRun.java:
```diff
@@ -439,6 +440,61 @@ public void testCustomActionDatasetAccess() throws Exception {
+  public void testWorkflowLocalDatasets() throws Exception {
+    ApplicationManager applicationManager = deployApplication(testSpace, WorkflowAppWithLocalDatasets.class);
+    final WorkflowManager wfmanager = applicationManager.getWorkflowManager("WorkflowWithLocalDatasets");
+    File waitFile = new File(tmpFolder.newFolder() + "/wait.file");
```
Use the `File(File parent, String path)` constructor.

chtyim commented:
Did a quick pass. Please address those comments first.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java:
```diff
@@ -101,6 +103,11 @@ public BasicMapReduceContext(Program program,
+    if (dsFramework instanceof WorkflowDatasetFramework) {
+      localDatasetNameMapping = ((WorkflowDatasetFramework) dsFramework).getDatasetNameMapping();
+    } else {
+      localDatasetNameMapping = null;
```
Better not to use null, use empty map instead.

sagarkapare replied:
Updated code to use the empty map instead.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java:
```diff
@@ -173,6 +176,26 @@ public WorkflowToken getWorkflowToken() {
+  public Map<String, String> getLocalDatasetNameMapping() {
+    String mappingJson = hConf.get(HCONF_ATTR_WORKFLOW_LOCAL_DATASET_NAME_MAPPING);
+    if (mappingJson == null) {
+      return null;
```
As mentioned above, better use empty map instead of null.

sagarkapare replied:
Using empty map.

sagarkapare commented:
@chtyim Addressed comments. Will add the Spark and runtime argument related changes in the separate PR. Please take a look again. Thanks.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java:
```diff
@@ -87,7 +90,7 @@
   private final LocationFactory locationFactory;
   private final MetricsCollectionService metricsCollectionService;
-  private final DatasetFramework datasetFramework;
+  private DatasetFramework datasetFramework;
```
Better leave this field final and have it representing the original datasetFramework that get injected through constructor. In the run() method, just have a local variable for the optionally new ds framework:

```java
DatasetFramework workflowDSFramework = datasetFramework;
if (arguments.hasOption(...)) {
  workflowDSFramework = new WorkflowDatasetFramework(...);
}
```

The reason is if anyone uses the datasetFramework field before the rewrite, it will be wrong.

sagarkapare replied:
Fixed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java:
```diff
@@ -172,11 +174,15 @@ public BasicMapReduceTaskContext load(ContextCacheKey key) throws Exception {
 
+        Map<String, String> localDatasetNameMapping = contextConfig.getLocalDatasetNameMapping();
+        DatasetFramework dsFramework = localDatasetNameMapping == null ? datasetFramework
+          : new WorkflowDatasetFramework(datasetFramework, localDatasetNameMapping);
```
This seems wrong as inside MapReduce it shouldn't be aware of workflow or local dataset at all. Seems like it should be done at the guice level.

sagarkapare commented:
Updated PR to use the commit from #5258. 
Following refactoring is done:
1. Added class `WorkflowProgramRunnerFactory` which implements the `ProgramRunnerFactory`.
2. This class is responsible for creating instances of the `MapReduceProgramRunner` and `SparkProgramRunner`.
3. `WorkflowDriver` instantiates the `WorkflowProgramRunnerFactory` and passes along to the `ProgramWorkflowRunnerFactory`

sagarkapare commented:
@chtyim Please take a look again. Thanks!

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -145,12 +151,35 @@
+                                                                         program, options);
+  }
+
+  private DatasetFramework createForWardingDatasetFramework(DatasetFramework datasetFramework, String runId) {
```
Better call it `createWorkflowDatasetFramework`.

sagarkapare replied:
Renamed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -145,12 +151,35 @@
-    this.datasetFramework = datasetFramework;
+    this.datasetFramework
+      = createForWardingDatasetFramework(datasetFramework,
+                                         options.getArguments().getOption(ProgramOptionConstants.RUN_ID));
```
It's better to use `runId.getId()`. Same applies to the creation of `loggingContext` above.

sagarkapare replied:
using `runId.getId()`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -427,14 +456,42 @@ private void executeCondition(ApplicationSpecification appSpec, WorkflowConditio
-    executeAll(workflowSpec.getNodes().iterator(), program.getApplicationSpecification(),
-               new InstantiatorFactory(false), program.getClassLoader(), token);
+    try {
+      createLocalDatasets();
```
Probably better to do create in the `startUp()` and delete in `shutDown`.

sagarkapare replied:
Moved the creation to `startUp` and deletion to `shutDown`

chtyim commented:
Just one comment left. Rest LGTM.

sagarkapare commented:
@chtyim Addressed comment. Running new build here - http://builds.cask.co/browse/CDAP-DUT3631-8. Will merge once the build passes.

chtyim commented:
LGTM

sagarkapare commented:
Build passed. Squashing and merging.
